src/Microsoft.AspNetCore.Blazor/RenderTree/RenderTreeDiffBuilder.cs

```diff
@@ -597,9 +597,15 @@ private static void InitializeNewComponentFrame(ref DiffContext diffContext, int
 
         private static void InitializeNewAttributeFrame(ref DiffContext diffContext, ref RenderTreeFrame newFrame)
         {
-            if (newFrame.AttributeValue is UIEventHandler)
+            if (newFrame.AttributeValue is MulticastDelegate @delegate &&
+                @delegate.Method != null)
             {
-                diffContext.Renderer.AssignEventHandlerId(ref newFrame);
+                var parameters = @delegate.Method.GetParameters();
+                if (parameters.Length == 1 && 
+                    typeof(UIEventArgs).IsAssignableFrom(parameters[0].ParameterType))
+                {
+                    diffContext.Renderer.AssignEventHandlerId(ref newFrame);
+                }
```

Comment:
> I assume we only want to 'register' delegates that can receive dom events

I'm not sure that's so important. Are there many likely scenarios where someone tries to use `onsomething=@myDelegate` where `myDelegate` isn't a `UIEventHandler`? Even if someone does that and we assign an event handler ID that never gets used, that doesn't seem like a big problem.

I'd favor optimizing for the common case (where the delegate is a `UIEventHandler`) even if in the uncommon case we do something a little less inefficient.

Reply:
We could end up doing something like "if it's on a component, leave it alone; if it's on an HTML element, then assign IDs to all delegate values" since there's no use case for delegate-typed HTML attribute values other than event handlers. I know we don't have the "component or not" info directly available here but could find some way to make it available (possibly as an extra param from the caller, or possible as extra info on the `RenderTreeFrame` itself).